### PR TITLE
refactor with max frequency per row

### PR DESCRIPTION
**Pull request by JesperDramsch**
Refactor of column names to make code more legible and easily changeable. Corrected column number for deletion days.

Add Column for max update frequency per row as suggested in #21 

This PR only works with the following sheet:
image

### CODE REVIEW

**Comment by JesperDramsch** [comment]
This way the update frequency for the sheet can be a high number like every hour, but each row is only updated according to column 3.

Works without values in column 3 and 4.

File: sheetScript.gs [hunk 1]
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
...
-        Logger.log("Channel with id "+channelIds[i]+" has no new videos")
-      } else {
-        [].push.apply(newVideoIds, videoIds);
-      }
+	
+	// Check if it's time to update already
+	var freqDate = new Date(lastTimestamp);
+	var dateDiff = Date.now() - freqDate;
+	var nextTime = sheet.getRange(iRow + 1, reservedColumnFrequency).getValue()  * 36e5;
+	if (dateDiff <= nextTime && nextTime) {
```
Comment (victorjzsun) [inline sheetScript.gs]: Check nextTime first before comparison with dateDiff

File: sheetScript.gs [hunk 2]
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
...
+		Logger.log("Skipped: Not time yet");
+	} else {
+	    /// ...get channels...
+	    var channelIds = [];
+	    var playlistIds = [];
+	    Logger.log("Row: " + (iRow+1))
```
Comment (victorjzsun) [inline sheetScript.gs]: Move this line higher up so that user knows which row was skipped from log on line 87

File: sheetScript.gs [hunk 3]
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
...
-    for (var i = 0; i < channelIds.length; i++) {
-      var videoIds = getVideoIdsWithLessQueries(channelIds[i], lastTimestamp)
-      if (!videoIds || typeof(videoIds) !== "object") addError("Failed to get videos with channel id "+channelIds[i])
-      else if (debugFlag_logWhenNoNewVideosFound && videoIds.length === 0) {
-        Logger.log("Channel with id "+channelIds[i]+" has no new videos")
-      } else {
-        [].push.apply(newVideoIds, videoIds);
-      }
+	
+	// Check if it's time to update already
```
Comment (victorjzsun) [inline sheetScript.gs]: Convert tabs to spaces for consistency

File: sheetScript.gs (hunk 1 above)
Comment (JesperDramsch) [reply sheetScript.gs]: Why?

File: sheetScript.gs [hunk 4]
```diff
@@ -25,12 +34,10 @@ function doGet(e) {
 }
 
 function findNextRow(data) { // Finds the row with the earliest last updated time
-  var reservedTimestampColumn = 1;
-  var reservedTableRows = 3; // Start of the range of the PlaylistID+ChannelID data
   var minTimestamp = data.slice(reservedTableRows).reduce(
     function (min, row, index) {
-      if (row[reservedTimestampColumn].length != 0 && row[reservedTimestampColumn] < min[1]) {
-        return [index, row[reservedTimestampColumn]]
+      if (row[reservedColumnPlaylist].length != 0 && row[reservedColumnPlaylist] < min[1]) {
```
Comment (victorjzsun) [inline sheetScript.gs]: Wouldn't `reservedColumnTimestamp - 1` be more readable?

File: sheetScript.gs [hunk 5]
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
...
+	    if (!errorflag) {
+	      // ...add videos to playlist...
+	      if (!debugFlag_dontUpdatePlaylists) {
+	        addVideosToPlaylist(playlistId, newVideoIds);
+	      } else {
+	        addError("Don't Update Playlists debug flag is set");
+	      }
+	      
+	      /// ...delete old vidoes in playlist
+	      var daysBack = data[iRow][reservedColumnDeleteDays];
```
Comment (victorjzsun) [inline sheetScript.gs]: One-off error since this is array index?

File: sheetScript.gs (hunk 4 above)
Comment (JesperDramsch) [reply sheetScript.gs]: That assumes you'll never change the order of the columns again. Considering that the code contained multiple mistakes with constant offsets (reservedTimestampColumn +1 +2 +3)  I recommend making these explicit at the beginning of the script.

File: sheetScript.gs (hunk 2 above)
Comment (JesperDramsch) [reply sheetScript.gs]: Good point.

File: sheetScript.gs (hunk 3 above)
Comment (JesperDramsch) [reply sheetScript.gs]: I was wondering about this. Done

File: sheetScript.gs (hunk 5 above)
Comment (JesperDramsch) [reply sheetScript.gs]: In the Row or Column?

File: sheetScript.gs (hunk 1 above)
Comment (victorjzsun) [reply sheetScript.gs]: You're checking if nextTime exists (isn't an empty string) here right? Wouldn't it be better to check for existence before comparing it with dateDiff?

Comment (JesperDramsch) [reply sheetScript.gs]: I'll change it, but I don't think it matters, as an empty cell evaluates to 0.

File: sheetScript.gs (hunk 5 above)
Comment (victorjzsun) [reply sheetScript.gs]: column

Comment (JesperDramsch) [reply sheetScript.gs]: There was an off by one error before, that I corrected. It was still pointing to the old second column.

Comment (JesperDramsch) [reply sheetScript.gs]: Or am I mixing things up here?

File: sheetScript.gs (hunk 4 above)
Comment (victorjzsun) [reply sheetScript.gs]: Do whatever you think makes sense, and allows future developers to understand what's going on.

File: sheetScript.gs (hunk 5 above)
Comment (victorjzsun) [reply sheetScript.gs]: All of your constants are sheet column indexes, and any data array index should be `<constant> - 1` (like the `reservedColumnTimestamp - 1` = `reservedColumnPlaylist` indexing in findNextRow)

Comment (JesperDramsch) [reply sheetScript.gs]: I think you're right I apologize. Let me try and fix it.

**Comment by JesperDramsch** [comment]
Okay, I changed the `getRange` codes to a call of the data array and gave an explanation in the comments with the variable to make it easier for new devs. Changed everything to the correct 0-index (sorry again). 

Now the only "off by X" correction has to be made when writing to the sheet, which is explained in the sheet.

Fixed a bug with a variable name `playlistId` vs `plalistID` that came up from time to time.

Addressed all commens and implemented those changes.

**Comment by JesperDramsch** [comment]
Thanks for the thorough review!

File: sheetScript.gs [hunk 6]
```diff
@@ -63,94 +70,101 @@ function updatePlaylists(sheet) {
...
-      } else {
-        [].push.apply(newVideoIds, videoIds);
+  
+  // Check if it's time to update already
+  var freqDate = new Date(lastTimestamp);
+  var dateDiff = Date.now() - freqDate;
+  var nextTime = data[iRow][reservedColumnFrequency]  * MILLIS_PER_HOUR;
+  if (nextTime && dateDiff <= nextTime) {
+    Logger.log("Skipped: Not time yet");
+  } else {
```
Comment (victorjzsun) [inline sheetScript.gs]: These lines should be 6-space indented for consistency with code above and below

File: sheetScript.gs [hunk 7]
```diff
@@ -11,6 +11,17 @@ var debugFlag_dontUpdateTimestamp = false;
 var debugFlag_dontUpdatePlaylists = false;
 var debugFlag_logWhenNoNewVideosFound = false;
 
+
+// Define reserved Rows and Columns (zero-based)
+var reservedTableRows = 2;        // Start of the range of the PlaylistID+ChannelID data
+var reservedTableColumns = 4;     // Start of the range of the ChannelID data
```
Comment (victorjzsun) [inline sheetScript.gs]: These were zero-based before, so should stay as
```
var reservedTableRows = 3;        // Start of the range of the PlaylistID+ChannelID data
var reservedTableColumns = 5;     // Start of the range of the ChannelID data
```
We want reservedTableRows to be row 4 of the sheet, and reservedTable Columns to be column F of the sheet, the start of user data

**Comment by JesperDramsch** [comment]
@victorjzsun All points addressed.

**Pull request merged**